File: shelf/src/test/java/com/toddway/shelf/CacheOrNewTests.java
package com.toddway.shelf;

import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.util.List;

import rx.Observable;
import rx.observers.TestSubscriber;

import static org.junit.Assert.assertEquals;

/**
 * Created by tway on 3/12/16.
 */
public class CacheOrNewTests {

    ShelfItem item;
    TestSubscriber<String> subscriber;
    Shelf shelf;
    String newValue;
    String cacheValue;

    @Before
    public void beforeEach() {
        shelf = new Shelf(new File("/tmp"));
        item = shelf.item("string");
        subscriber = new TestSubscriber<>();
        newValue = "new value";
        cacheValue = "cache value";
    }

    private void givenNoCache() {
        item.clear();
    }

    private void givenValidCache() {
        item.lifetime(5000).put(cacheValue);
    }

    private void givenInvalidCache() {
        item.lifetime(0).put(cacheValue);
    }

    private void givenNoNew() {
        newValue = null;
    }

    private void printValues(List<String> values) {
        for (String value : values) System.out.println("value: " + value);
    }

    private void whenCacheOrNewSubscription() {
        Observable.just(newValue)
                .compose(item.cacheOrNew(String.class))
                .subscribe(subscriber);
    }

    @Test
    public void testNoCache() {
        givenNoCache();
        whenCacheOrNewSubscription();

        subscriber.assertValues(newValue);
        assertEquals(item.get(String.class), newValue);
        printValues(subscriber.getOnNextEvents());
    }


    @Test
    public void testValidCache() {
        givenValidCache();
        whenCacheOrNewSubscription();

        subscriber.assertValues(cacheValue);
        subscriber.assertNoErrors();
        assertEquals(item.get(String.class), cacheValue);
        printValues(subscriber.getOnNextEvents());
    }

    @Test
    public void testInvalidCache() {
        givenInvalidCache();
        whenCacheOrNewSubscription();

        subscriber.assertValues(newValue);
        assertEquals(item.get(String.class), newValue);
        printValues(subscriber.getOnNextEvents());
    }

    @Test public void testNoCacheAndNoNew() {
        givenNoCache();
        givenNoNew();
        whenCacheOrNewSubscription();

        subscriber.assertNoValues();
        assertEquals(item.get(String.class), null);
        printValues(subscriber.getOnNextEvents());
    }

    @Test public void  testNoNewAndInvalidCache() {
        givenNoNew();
        givenInvalidCache();
        whenCacheOrNewSubscription();

        subscriber.assertNoValues();
        assertEquals(item.get(String.class), cacheValue);
        printValues(subscriber.getOnNextEvents());
    }
}

File: shelf/src/test/java/com/toddway/shelf/CacheOrNewTests.kt
package com.toddway.shelf

import com.toddway.shelf.rx.cacheOrNew
import io.reactivex.Single
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.IOException

/**
 * Created by tway on 3/12/16.
 * updated nschwermann 12/21/17
 */
class CacheOrNewTests : BaseTest(){

    private val noElementError : IOException = IOException()

    private fun whenCacheOrNewSubscription() {

        Single.defer { newValue?.let { Single.just(it) } ?: Single.error(noElementError) }
                .cacheOrNew(item, String::class.java)
                .subscribe(subscriber)
    }

    @Test
    fun testNoCache() {
        givenNoCache()
        whenCacheOrNewSubscription()

        subscriber.assertValues(newValue!!)
        assertEquals(item[String::class.java], newValue)
        printValues(subscriber.values())
    }


    @Test
    fun testValidCache() {
        givenValidCache()
        whenCacheOrNewSubscription()

        subscriber.assertValues(cacheValue)
        subscriber.assertNoErrors()
        assertEquals(item[String::class.java], cacheValue)
        printValues(subscriber.values())
    }

    @Test
    fun testInvalidCache() {
        givenInvalidCache()
        whenCacheOrNewSubscription()

        subscriber.assertValues(newValue!!)
        assertEquals(item[String::class.java], newValue)
        printValues(subscriber.values())
    }

    @Test
    fun testNoCacheAndNewError() {
        givenNoCache()
        givenNoNew()
        whenCacheOrNewSubscription()

        subscriber.assertNoValues()
        subscriber.assertError(noElementError)
    }

    @Test
    fun testNewErrorAndInvalidCache() {
        givenNoNew()
        givenInvalidCache()
        whenCacheOrNewSubscription()

        subscriber.assertNoValues()
        subscriber.assertError(noElementError)
        assertEquals(item[String::class.java], cacheValue)
        printValues(subscriber.values())
    }
}
